draft/XXXX-robust-api/README.md

```diff
@@ -0,0 +1,146 @@
+---
+RFC: XXXX
+Author: Jacob Nesbitt
+Created: 2020-06-29
+Created: 2020-12-12
+Last-Modified: 2020-12-12
+---
+
+# RFC XXXX: RFC Template
+
+This RFC outlines a robust way to handle table/graph creation and upload, allowing for arbitrarily large datasets to be uploaded in any of the supported formats.
+
+## Motivation
+
+As the size of our datasets has grown, our limited approach to handling data upload and table/graph creation has proved to be inefficient. The need has arose for an API that can handle larger data, with more operations. The proposals outlined below will make that possible.
+
+# Proposal
+
+## Upload
+Currently, our upload system deals with upload, data processing, and table creation, all at once and in the same request. Since we deploy on heroku, and we need to keep our requests under 30 seconds (as we should regardless), this severely limits both the size of the data we can upload. In order to address this, we can make use of Amazon S3. The new upload process would follow this rough path:
+
+
+### Create an upload
+
+`POST /api/workspaces/{workspace}/upload/create`
+
+Create an upload, returning an `Upload` object that (among other things) will contain a list of pre-signed URLs, that are to be used for uploading each part of the file to S3. **Note:** This means that the amount of parts / size of each part is determined by the server. The parameters for this call will include the following
+
+* `dataType` - The data type of the file to upload, which must be one of the supported (or installed, if using a plugin system) data types. E.g. `csv`, `d3_json`, `nested_json`, `newick`, etc. If this parameter does not match one of the supported data types, a `4XX` error is thrown.
+* `size` - The size of the file, in bytes. This is needed to determine the number of pre-signed URLs to create for this upload object.
+
+The response will be an upload object, which will have the following structure
+
+```
+{
+  "_id": string (uuid),
+  "user": string (uuid),
+  "started": string (datetime),
+  "finished": string (datetime) | null,
+  "dataType": string (must be one of supported data types),
+  "objectType": "table" | "graph",
+  "status": "pending" | "processing" | "finished",
+  "uploadId": string (uuid),
+  "uploadPartUrls": [
+    "<signed url for first part>",
+    "<signed url for second part>",
+    ...
+  ],
+  "uploadPartTags": [ // This will be an empty list initially
+    "<ETag from the first uploaded part>",
+    "<ETag from the second uploaded part>",
+    ...
+  ],
+  "metadata": {
+    // metadata like type information goes here
+  }
+}
```

Comment:
We might want to compute a hash so that we can check if the upload was already started. If it was, we can resume instead of starting over again

Reply:
Resumption wouldn't be through this `POST` endpoint, though. The `_id` of the response would be used for that, I think. Upload interruption/resumption would have to be handled through a specialized system that would have a UI in the client.

I agree that a hash would be useful in verifying that a resumed upload is occurring with the same file that was interrupted before, though. On the other hand, I'm sure there are best practices for this kind of thing and we should find out what those are.

Reply:
I'm a bit confused. The only reason I could see us wanting to store a hash here is to hash the original file before uploading, which could then be used if the client is refreshed or the connection is broken, to ensure that the file they select to resume the upload exactly matches the original file.


In my opinion this is overkill, and I don't think we should worry about this to start. Adding a `filename` and `size` field to this metadata should be sufficient, and would allow the user to identify which file it is they need to select to resume upload.

As far as how the user would know which uploads they have in progress, there should be an endpoint of the form

```
GET /api/users/uploads
```

which would return a list of the uploads that the currently logged in user initialized. This is currently lacking from the RFC, so I'll ad that back in.

Reply:
> I'm a bit confused. The only reason I could see us wanting to store a hash here is to hash the original file before uploading, which could then be used if the client is refreshed or the connection is broken, to ensure that the file they select to resume the upload exactly matches the original file.
> 
> In my opinion this is overkill, and I don't think we should worry about this to start. Adding a `filename` and `size` field to this metadata should be sufficient, and would allow the user to identify which file it is they need to select to resume upload.

No, you got it, this is just what I meant. The filename+size combo is a sort of hash, just one that doesn't prevent the file contents themselves from changing, but also one that doesn't require being stored somewhere.

I don't believe it's overkill to record the hash to prevent crazy (though rare) situations that would lead to corrupted data, but I agree with you that it's something we really don't need to do now. I'll file an issue so we don't forget about it.

> 
> As far as how the user would know which uploads they have in progress, there should be an endpoint of the form
> 
> ```
> GET /api/users/uploads
> ```
> 
> which would return a list of the uploads that the currently logged in user initialized. This is currently lacking from the RFC, so I'll ad that back in.

I think this `GET` endpoint should carry the same form as the `POST` endpoint you've proposed in this RFC. That one is targeted to workspaces, so perhaps there ought to simply be a `/api/uploads` section, and the model can record what workspace each upload is targeted to. Now that I'm thinking about it, I'm not sure the upload model even needs the workspace info, since the finalize call could be informed about which workspace to target the download to.